uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-request-popup-details-tab.component.js

```diff
@@ -141,52 +136,138 @@ define([
+      })
+      .then(!vm.isMode('create') && setDatepickerBoundaries)
       .then(initOriginalOpeningBalance)
+      .then(setOpeningBalance)
       .then(function () {
         return $q.all([
-          setDaySelectionMode(),
+          setDaysSelectionMode(),
           initBalanceChange()
         ]);
       })
-      .then(initTimeAndDateInputsWatchers)
+      .then(function () {
+        if (!vm.isMode('view') && !isLeaveType('toil')) {
+          initTimeAndDateInputsWatchers();
+        }
+      })
       .catch(handleError)
       .finally(function () {
         vm.loading.tab = false;
       });
     }());
 
     /**
-     * Calculate change in balance, it updates local balance variables.
+     * Calculates closing balance which is opening balance minus change amount.
+     * We use "+" operation since the change amount is negative.
+     */
+    function calculateClosingBalance () {
+      vm.balance.closing = vm.balance.opening + vm.balance.change.amount;
+    }
+
+    /**
+     * Converts given date to javascript date as expected by uib-datepicker
      *
-     * @return {Promise} empty promise if all required params are not set
-     *   otherwise promise from server
+     * @param {String} date from server
+     * @return {Date}
      */
-    function calculateBalanceChange () {
-      vm._setDateAndTypes();
-      vm._toggleBalance();
+    function convertDateFormatFromServer (date) {
+      return moment(date, sharedSettings.serverDateFormat).toDate();
+    }
 
-      if (!vm._canCalculateChange()) { return $q.resolve(); }
+    /**
+     * Converts given date to server format
+     *
+     * @param {Date} date
+     * @return {String} date converted to server format
+     */
+    function convertDateToServerFormat (date) {
+      return moment(date).format(sharedSettings.serverDateFormat);
+    }
 
-      vm.loading.showBalanceChange = true;
+    /**
+     * Sets dates (with time in case of "hours" Absence Type) from UI to request
+     *
+     * @param  {String} dateType from|to
+     * @return {Promise}
+     */
+    function dateChangeHandler (dateType) {
+      var absencePeriodChanged = isAbsencePeriodChanged(dateType);
 
-      return vm.request.calculateBalanceChange(vm.selectedAbsenceType.calculation_unit_name)
-        .then(setBalanceChange)
+      setDatepickerBoundaries();
+      (dateType === 'from') && resetToDateIfGreaterThanFromDate();
+      resetUIDayTypesTimeAndDeductions(dateType);
+
+      vm.uiOptions.times[dateType].loading = true;
+      vm.loading[dateType + 'DayTypes'] = true;
+      vm.uiOptions.times[dateType].disabled = true;
+
+      return $q.resolve()
+        .then(vm.onDateChangeExtended)
+        .then(function () {
+          if (absencePeriodChanged) {
+            vm.period = getAbsencePeriod(dateType);
+          }
+
+          if (!vm.period.id) {
+            return $q.reject('Please change date as it is not in any absence period');
+          }
+        })
+        .then(loadCalendar)
```

Comment:
why are you loading the calendar (with subsequent `$q.all()`) even if the absence period didn't change?

Reply:
Done.